File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/task/MoleTask.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.core.implementation.task;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedList;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.core.model.data.IDataSet;
import org.openmole.core.model.execution.IProgress;
import org.openmole.core.model.job.IContext;
import org.openmole.commons.exception.MultipleException;
import org.openmole.commons.aspect.caching.SoftCachable;
import org.openmole.commons.aspect.eventdispatcher.IObjectChangedSynchronousListenerWithArgs;
import org.openmole.commons.tools.service.Priority;
import org.openmole.core.implementation.data.DataSet;
import org.openmole.core.implementation.internal.Activator;
import org.openmole.core.implementation.job.Context;
import org.openmole.core.implementation.job.SynchronizedContext;
import org.openmole.core.implementation.mole.MoleExecution;
import org.openmole.core.model.data.IData;
import org.openmole.core.model.data.IPrototype;
import org.openmole.core.model.job.IMoleJob;
import org.openmole.core.model.job.State;
import org.openmole.core.model.mole.IMole;
import org.openmole.core.model.mole.IMoleExecution;
import org.openmole.core.model.task.IMoleTask;

public class MoleTask extends Task implements IMoleTask {

    class ExceptionLister implements IObjectChangedSynchronousListenerWithArgs<IMoleExecution> {

        final Collection<Throwable> throwables = Collections.synchronizedCollection(new LinkedList<Throwable>());

        @Override
        public void objectChanged(IMoleExecution t, Object[] os) throws InternalProcessingError, UserBadDataError {
            IMoleJob moleJob = (IMoleJob) os[0];

            if (moleJob.getState() == State.FAILED) {
                Throwable exception = moleJob.getContext().getValue(GenericTask.Exception.getPrototype());
                throwables.add(exception);
            }
        }

        public Collection<Throwable> getThrowables() {
            return throwables;
        }
    }
    IMole mole;

    public MoleTask(String name, IMole workflow)
            throws UserBadDataError, InternalProcessingError {
        super(name);
        this.mole = workflow;
    }

    @Override
    protected void process(IContext global, IContext context, IProgress progress) throws UserBadDataError, InternalProcessingError, InterruptedException {

        IContext globalContext = new SynchronizedContext();
        IContext firstTaskContext = new Context();

        for (IData<?> input : getInput()) {
            if (!input.getMode().isOptional() || input.getMode().isOptional() && context.contains(input.getPrototype())) {
                IPrototype p = input.getPrototype();
                firstTaskContext.putVariable(p, context.getValue(p));
            }
        }

        IMoleExecution execution = new MoleExecution(mole, globalContext, firstTaskContext);

        ExceptionLister exceptionLister = new ExceptionLister();
        Activator.getEventDispatcher().registerListener(execution, Priority.NORMAL.getValue(), exceptionLister, IMoleExecution.oneJobFinished);

        execution.start();
        execution.waitUntilEnded();

        for (IData<?> output : getUserOutput()) {
            IPrototype p = output.getPrototype();
            if (globalContext.contains(p)) {
                context.putVariable(p, globalContext.getValue(p));
            }
        }

        Collection<Throwable> exceptions = exceptionLister.getThrowables();

        if (!exceptions.isEmpty()) {
            context.putVariable(GenericTask.Exception.getPrototype(), new MultipleException(exceptions));
        }
    }

    @Override
    public IMole getMole() {
        return mole;
    }

    @SoftCachable
    @Override
    public IDataSet getInput() throws InternalProcessingError, UserBadDataError {
        return new DataSet(super.getInput(), getMole().getRoot().getAssignedTask().getInput());
    }
}

File: openmole/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/task/MoleTask.scala
/*
 * Copyright (C) 2010 Romain Reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.task

import org.openmole.misc.tools.service.Random
import org.openmole.misc.eventdispatcher._
import org.openmole.misc.exception._
import org.openmole.misc.tools.obj.ClassUtils._
import org.openmole.core.implementation.puzzle._
import org.openmole.core.implementation.data._
import org.openmole.core.implementation.mole._
import org.openmole.core.implementation.mole.Capsule._
import org.openmole.core.model.mole._
import org.openmole.core.model.data._
import org.openmole.core.model.mole._
import org.openmole.core.model.task._

object MoleTask {

  def apply(name: String, puzzle: Puzzle)(implicit plugins: PluginSet): TaskBuilder =
    apply(name, new Mole(puzzle.first), puzzle.lasts.head)

  def apply(name: String, mole: IMole, last: ICapsule)(implicit plugins: PluginSet) = {
    new TaskBuilder { builder ⇒
      def toTask = new MoleTask(name, mole, last) {
        val inputs = builder.inputs + mole.root.inputs
        val outputs = builder.outputs + last.outputs
        val parameters = builder.parameters
      }
    }
  }

}

sealed abstract class MoleTask(
    val name: String,
    val mole: IMole,
    val last: ICapsule)(implicit val plugins: PluginSet) extends Task with IMoleTask {

  class ResultGathering extends EventListener[IMoleExecution] {
    var lastContext: Option[Context] = None
    var exceptions: List[Throwable] = List.empty

    override def triggered(obj: IMoleExecution, ev: Event[IMoleExecution]) = synchronized {
      ev match {
        case ev: IMoleExecution.JobInCapsuleFinished ⇒
          if (ev.capsule == last) lastContext = Some(ev.moleJob.context)
        case ev: IMoleExecution.ExceptionRaised ⇒
          exceptions ::= ev.exception
        case _ ⇒
      }
    }
  }

  override protected def process(context: Context) = {
    val firstTaskContext = inputs.foldLeft(List.empty[Variable[_]]) {
      (acc, input) ⇒
        if (!(input.mode is optional) || ((input.mode is optional) && context.contains(input.prototype)))
          context.variable(input.prototype).getOrElse(throw new InternalProcessingError("Bug: variable not found.")) :: acc
        else acc
    }.toContext

    val execution = new MoleExecution(mole, rng = Random.newRNG(context.valueOrException(Task.openMOLESeed)))
    val resultGathering = new ResultGathering

    EventDispatcher.listen(execution: IMoleExecution, resultGathering, classOf[IMoleExecution.JobInCapsuleFinished])
    EventDispatcher.listen(execution: IMoleExecution, resultGathering, classOf[IMoleExecution.ExceptionRaised])

    execution.start(firstTaskContext)
    execution.waitUntilEnded

    if (!resultGathering.exceptions.isEmpty) throw new MultipleException(resultGathering.exceptions.reverse)

    context + resultGathering.lastContext.getOrElse(throw new UserBadDataError("Last capsule " + last + " has never been executed."))
  }

}
